demos/tutorials/RandomVariables.py

```diff
@@ -0,0 +1,386 @@
+# the ``dist`` attribute.
+
+print(x.dist)
+
+
+# %%
+######################################################################
+# Random variables act differently than regular Python
+# variables in the following important ways:
+# 
+# 1. They are not immediately evaluated when defined. Instead, they are
+#    evaluated in a lazy fashion in the context needed.
+# 2. They contain at least one probability distribution (single or multivariate)
+#    instead of a single value or vector of values.
+# 
+# Despite these differences, random variables can be used in the same way
+# as regular variables in most cases, such as performing algebraic operations.
+# 
+# For example suppose we want to define a new variable :math:`y` as
+# follows:
+# 
+# .. math::
+# 
+# 
+#    y = (x + 10)^2
+# 
+# This can be done in CUQIpy as follows:
+# 
+
+y = (x + 10)**2
+
+
+# %%
+######################################################################
+# Calling print on ``y`` reveals that it is a random variable which has
+# recorded the operations performed on it and maintains a reference to the
+# random variable ``x`` and its underlying distribution.
+#
+# Here it is worth it to note that performing algebraic operations on
+# directly on the distributions, e.g. (X+10)**2, is not possible.
+# 
+
+print(y)
+
+# %%
+######################################################################
+# Transformed random variables
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# The random variable ``y`` defined above is an example of a transformed
+# random variable. Transformed random variables are random variables which
+# are defined by applying algebraic operations to other random variables.
+# 
+# In CUQIpy, it is possible to determine if a random variable is
+# transformed by checking the ``is_transformed`` attribute.
+#
+
+print(y.is_transformed)
+
+
+# %%
+######################################################################
+# Evaluating random variables
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# 
+# Currently, a two modes of evaluation are supported for all
+# (simple and transformed) random variables:
+# 
+# 1. Sampling
+# 2. Direct evaluation
+# 
+# Sampling
+# ^^^^^^^^
+# 
+# Sampling works by drawing a random sample from the underlying probability
+# distribution of the random variable. This is done using the ``sample``
+# method.
+# 
+# This is possible because the random variable knows of any
+# transformations performed on it and can use this information to
+# transform the sample drawn from the original distribution.
+# 
+# This means for example that we can both draw a sample from :math:`x` and
+# :math:`y` using the same syntax, which also matches the syntax used for
+# sampling from distributions.
+# 
+
+# Draws a sample from x ~ Gaussian(0,1)
+print(x.sample()) 
+
+# Draws a sample from y = (x+10)^2, x ~ Gaussian(0,1)
+print(y.sample()) 
+
+
+# %%
+######################################################################
+# Direct evaluation
```

Comment:
May be we can clarify to the user here or in some other place in the tutorial that using () operator means different things when we talk about X and x, X(1) is logpdf and X.rv(1) is the evaluation as described here.